### PR TITLE
TTRSS-Backend startup optimization

### PR DESCRIPTION
**Pull request by noctux**
The primary objective of this pull request is to speed up newsboat's startup when using the TTRSS-backend. Up to now, the TTRSS-backend's `get_subscribed_urls` used one request per feed category to collect its feedlist. Using the api-method "getFeeds", the number of these requests can be reduced to a single request. The old behaviour is preserved for servers with an api level < 2. This is the first step in adressing #16.

Beyond this change, this pullrequest contains a couple of minor fixes I've accumulated in the process, within separate commits:
- properly handling the quotation mark (") character within json payloads via escaping
- eliminating the markreadthread class by using c++11-lambdas
- Using libcurl's url-escaping function instead of our own hacky implementation (which only handles "?" and "&")
- experimenting with the c++-json library [nlohman/json](https://github.com/nlohmann/json), as an initial attempt at #103.

Sorry for this mixup of different issues into one pullrequest. I've been dogfooding the commits for some time, without any apparent issues (running TTRSS v17.4, b0113ad).
Please let me know if you've got any questions.

### CODE REVIEW

**Comment by noctux** [comment]
Before merging: We might want to have the json.hpp library as a submodule instead of including it, what do you think?

**Comment by noctux** [comment]
Sorry, my bad, forgot to adapt `mk/newsboat.deps` after removing the `markreadthread` class and files.

**Comment by Minoru** [comment]
Thank you! This might take some time to review, so I'm going to answer the easier question now, and the review should land sometime during the week.

> Before merging: We might want to have the json.hpp library as a submodule instead of including it, what do you think?

I fail to see any benefit in submodules here, but see a downside: they burden clones and checkouts with extra CLI arguments and hassle. Since we're only interested in library's releases, which are less frequent than commits to their repo, I think it's okay to just copy the single file we need. Same logic applies to Catch. Do you see any benefits of submodules that I don't see? (Apart from affecting our test coverage so much :)

**Comment by noctux** [comment]
> the review should land sometime during the week.

Take your time, we aren't on the run :)

> Do you see any benefits of submodules that I don't see? (Apart from affecting our test coverage so much :)

Well, the benefit I see is better traceability for external viewers: the submodule is explicit, demonstrating that this is not our code in particular (could although be fixed by a "thirdparty" subfolder), and it is more apparent which version we are using (I mean: what version have I checked into this commit? Is it a release? One would hope. but can you be sure? How can you find out?). Depending on how we measure coverage, the potential improvement in the coverage stats is only a nice sideeffect :) Still, this is only a small aspect, so I've got no strong feelings there...

File: src/ttrss_api.cpp [hunk 1]
```diff
@@ -161,6 +161,35 @@ json ttrss_api::run_op(const std::string& op,
...
 
+tagged_feedurl ttrss_api::feed_from_json(const json& jfeed, std::function<std::vector<std::string>(const json&)> tagger) {
+	const int feed_id            = jfeed["id"];
+	const std::string feed_title = jfeed["title"];
+	const std::string feed_url   = jfeed["feed_url"];
+
+	std::vector<std::string> tags;
+	tags.push_back(std::string("~") + feed_title);
+
+	for(auto& tag : tagger(jfeed)) {
```
Comment (Minoru) [inline src/ttrss_api.cpp]: If you don't plan  `tagger` to ever return more than one tag, I'd rather see it replaced by a simple `std::string`, defaulted to an empty value. Current implementations of `tagger` can be written in-place in the loops from which you call `feed_from_json` (see comments below).

File: src/ttrss_api.cpp [hunk 2]
```diff
@@ -403,21 +417,13 @@ void ttrss_api::fetch_feeds_per_category(
...
-		if (!cat_name.is_null()) {
-			tags.push_back(cat_name.get<std::string>());
-		}
-
-		auto url = strprintf::fmt("%s#%d", feed_url, feed_id);
-		feeds.push_back(tagged_feedurl(url, tags));
-
-		// TODO: cache feed_id -> feed_url (or feed_url -> feed_id ?)
+		feeds.push_back(feed_from_json(feed, [&](const json&){
+				std::vector<std::string> tags;
```
Comment (Minoru) [inline src/ttrss_api.cpp]: This whole calculation doesn't depend on the feed, so you can move it out of the loop and replace this body with simple `return pre_calculated_vector`.

File: src/ttrss_api.cpp [hunk 3]
```diff
@@ -192,21 +214,13 @@ std::vector<tagged_feedurl> ttrss_api::get_subscribed_urls() {
...
-
-			if (cat_id > 0) {
-				tags.push_back(category_names[cat_id]);
-			}
-
-			auto url = strprintf::fmt("%s#%d", feed_url, feed_id);
-			feeds.push_back(tagged_feedurl(url, tags));
-			// TODO: cache feed_id -> feed_url (or feed_url -> feed_id ?)
+			feeds.push_back(feed_from_json(feed, [&](const json& f){
+					const int cat_id = f["cat_id"];
```
Comment (Minoru) [inline src/ttrss_api.cpp]: This body can be moved out of `push_back` into the body of the loop; then, you can pass `feed_from_json` the result. The code becomes more straightforward, and we avoid scaring people with lambdas :)

File: include/ttrss_api.h [hunk 4]
```diff
@@ -27,6 +27,8 @@ class ttrss_api : public remote_api {
 		void fetch_feeds_per_category(const nlohmann::json& cat, std::vector<tagged_feedurl>& feeds);
 		bool star_article(const std::string& guid, bool star);
 		bool publish_article(const std::string& guid, bool publish);
+		tagged_feedurl feed_from_json(const nlohmann::json& jfeed, std::function<std::vector<std::string>(const nlohmann::json&)> tagger);
```
[the remark this answers is not in the document]
Comment (noctux) [reply include/ttrss_api.h]: Browsing through recent commits, two tabs indent (relative to the return type indent) seems to be more common for this case in the codebase. Which one do you prefer?

File: src/ttrss_api.cpp (hunk 1 above)
Comment (noctux) [reply src/ttrss_api.cpp]: Based on your comment, I've implemented both variants (receiving a vector as well as a primitive string) as separate commits within this branch.
While I do not have immediate plans to expand the implementation by adding additional automatic tags, one should note that TTRSS itself has tags (which are currently disregarded), which would most likely be added via this codepath.
Looking at both implementations, I would argue that the vector variant is clean enough, while maintaining easier extendability - at a slight runtime cost for constructing the vector.

File: src/ttrss_api.cpp (hunk 2 above)
Comment (noctux) [reply src/ttrss_api.cpp]: Good catch! The lambda is now eliminated.

File: src/ttrss_api.cpp (hunk 3 above)
Comment (noctux) [reply src/ttrss_api.cpp]: Good catch, the lambda is now eliminated! I guess, too much functional programming leads to spurious visions of lambdas from time to time :)

File: src/ttrss_api.cpp [hunk 5]
```diff
@@ -161,6 +161,35 @@ json ttrss_api::run_op(const std::string& op,
 	return content;
 }
 
+tagged_feedurl ttrss_api::feed_from_json(const json& jfeed, std::function<std::vector<std::string>(const json&)> tagger) {
```
[the remark this answers is not in the document]
Comment (noctux) [reply src/ttrss_api.cpp]: Done

**Comment by noctux** [comment]
Thanks for your review! I fully agree with you, hopefully I've addressed all issues that you've brought up so far.

File: include/ttrss_api.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (Minoru) [reply include/ttrss_api.h]: Two is proper. Sorry for misguiding you!

File: src/ttrss_api.cpp [hunk 6]
```diff
@@ -161,6 +161,35 @@ json ttrss_api::run_op(const std::string& op,
...
+	std::vector<std::string> tags;
+	tags.push_back(std::string("~") + feed_title);
+
+	for(auto& tag : tagger(jfeed)) {
+		tags.push_back(tag);
+	}
+
+	auto url = strprintf::fmt("%s#%d", feed_url, feed_id);
+	return tagged_feedurl(url, tags);
+	// TODO: cache feed_id -> feed_url (or feed_url -> feed_id ?)
```
[the remark this answers is not in the document]
Comment (Minoru) [reply src/ttrss_api.cpp]: Ah, indeed! Sorry, I must've read the diff wrong and thought the comment is yours.

I'm in opposite position: I know a bit about our cache layer, but have no idea what's going on with TT-RSS API. Let that comment lay then, for now. :)

File: src/ttrss_api.cpp (hunk 1 above)
Comment (Minoru) [reply src/ttrss_api.cpp]: Agreed—vector version is clean enough.

File: src/ttrss_api.cpp (hunk 2 above)
Comment (Minoru) [reply src/ttrss_api.cpp]: The lambda is gone, but the calculation still happens inside the loop. The result is constant, so I don't see a reason not to move it out of the loop.

Comment (noctux) [reply src/ttrss_api.cpp]: And this is why you do code reviews :) You're absolutely right, should be fixed now.

**Comment by Minoru** [comment]
Thank you!

**Comment by Minoru** [comment]
By the way, would you like to be added as a contributor? You'll have commit access to the repo and will be able to fix small issues directly (though I'd ask you to still use PRs for larger things). You'll also be able to e.g. tag issues, if you're interested in that kind of thing.

**Comment by noctux** [comment]
Thanks for merging!

> By the way, would you like to be added as a contributor?

I'd be honored :) But rest assured, I'll still do PRs just so that I can later blame the reviewer (just kidding, I mean this review shows that another set of eyes is always a good idea...).

> You'll also be able to e.g. tag issues, if you're interested in that kind of thing.

You know, you're doing such a good job at bug triaging that I think I should better abstain from that well running machine :p

**Comment by Minoru** [comment]
Cool! Sent you an invite.

Well-running triage machine is in need of spare cogs; I'll re-try later I guess ;)

**Pull request merged**